chat-widget/src/components/livechatwidget/common/setPostChatContextAndLoadSurvey.ts

```diff
@@ -33,4 +42,205 @@ export const setPostChatContextAndLoadSurvey = async (chatSDK: any, dispatch: Di
+
+    // Check if Postchat already in progress
+    if (state.appStates.postChatWorkflowInProgress) {
+        // Handle case where chat is ended by customer
+        if (state.appStates.conversationEndedBy === ConversationEndEntity.Customer) {
+            await endChat(props, chatSDK, setAdapter, setWebChatStyles, dispatch, adapter, false, false, true);
+        }
+    }
+
+    // Conversation Details call required by customer as well as agent 
+    const conversationDetails = await getConversationDetailsCall(chatSDK);
+    // Start Postchat workflow
+    dispatch({ type: LiveChatWidgetActionType.SET_POST_CHAT_WORKFLOW_IN_PROGRESS, payload: true });
+
+    // Below logic checks if chat is ended by customer or agent or bot and handles them separately
+    if (state.appStates.conversationEndedBy === ConversationEndEntity.Customer) {
+        await postChatInitiatedByCustomer(props, chatSDK, setAdapter, setWebChatStyles, dispatch, adapter, state, postChatSurveyMode, conversationDetails);
+    } else if (state.appStates.conversationEndedBy === ConversationEndEntity.Agent) {
+        if (conversationDetails?.participantType === Constants.userParticipantTypeTag) { 
+            await postChatInitiatedByAgent(props, chatSDK, setAdapter, setWebChatStyles, dispatch, adapter, state, postChatSurveyMode);
+        } else if (conversationDetails?.participantType === Constants.botParticipantTypeTag) {
+            await postChatInitiatedByBot(props, chatSDK, setAdapter, setWebChatStyles, dispatch, adapter, state, postChatSurveyMode);
+        } else {
+            TelemetryHelper.logActionEvent(LogLevel.ERROR, {
+                Event: TelemetryEvent.AppStatesException,
+                ExceptionDetails: {
+                    exception: `ConversationDetails was not set correctly: conversationDetails = ${JSON.stringify(conversationDetails)}`
+                }
+            });
+            // Ending chat because something went wrong
+            await endChat(props, chatSDK, setAdapter, setWebChatStyles, dispatch, adapter, false, false, true);
+        }       
+    } else {
+        TelemetryHelper.logActionEvent(LogLevel.ERROR, {
+            Event: TelemetryEvent.AppStatesException,
+            ExceptionDetails: {
+                exception: `Conversation was Ended but App State was not set correctly: conversationEndedBy = ${state.appStates.conversationEndedBy}`
+            }
+        });
+        // Ending chat because something went wrong
+        await endChat(props, chatSDK, setAdapter, setWebChatStyles, dispatch, adapter, false, false, true);
+    }
+};
+
+// Function for link mode postchat workflow which is essentially same for both customer and agent
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const linkModePostChatWorkflow = (props: any, dispatch: Dispatch<ILiveChatWidgetAction>, setWebChatStyles: any) => {
+    TelemetryHelper.logActionEvent(LogLevel.INFO, {
+        Event: TelemetryEvent.LinkModePostChatWorkflowStarted
```

Comment:
In general, please check telemetry and make sure all events make sense and are not overflowing